File: algorithms/algorithms-cplusplus/coding_interviews_2/Question_0006_PrintListInReversedOrder.cc
#include <vector>

using namespace std;

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
 public:
  vector<int> printListFromTailToHead(ListNode* head) {      
    vector<int> res;
    help(head, res);
    return res;
  }

  void help(ListNode* head, vector<int>& res) {
    if (!head) {
      return;
    }
    help(head->next, res);
    res.push_back(head->val);
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/coding_interviews_2/Question_0006_PrintListInReversedOrder.java
package com.lanceolata.coding.algorithms.java.coding_interviews_2;

import java.util.ArrayList;

public class Question_0006_PrintListInReversedOrder {
    // Definition for singly-linked list.
    public static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public ArrayList<Integer> printListFromTailToHead(ListNode listNode) {
        ArrayList<Integer> res = new ArrayList<Integer>();
        help(listNode, res);
        return res;
    }

    private void help(ListNode listNode, ArrayList<Integer> res) {
        if (listNode == null) {
            return;
        }
        help(listNode.next, res);
        res.add(listNode.val);
    }
}
